File: merge-strings-alternately.cpp
class Solution {
public:
    string mergeAlternately(string word1, string word2) {
        std::string res;
        int len = word1.size() + word2.size();
        res.reserve(len);
        int i = 0, j = 0;

        while ((i == j) && !(i + j == len)) {
            if (i < word1.size()) {
                res += word1[i++];
            }
            if (j < word2.size()) {
                res += word2[j++];        
            }
        }

        for (; i < word1.size(); i++) {
            res += word1[i];
        }

        for (; j < word2.size(); j++) {
            res += word2[j];
        }

        return res;
    };
};

File: merge-strings-alternately.py
from collections import deque

class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        lst1, lst2 = deque(word1), deque(word2)
        res = []

        while lst1 and lst2:
            res.append(lst1.popleft())
            res.append(lst2.popleft())

        if lst1:
            res += lst1
        elif lst2:
            res += lst2

        return "".join(res)
